### PR TITLE
Fix task instance concurrency limit in a DAG affecting other DAGs

### PR DESCRIPTION
**Pull request by ephraimbuddy**
Currently, if a DAG with large tasks reaches a concurrency limit, task instances
in other dags will get stuck in scheduled state. The reason is that at each
scheduler loop, task instances of this large DAG are picked up for examination
despite that the DAG has reached the concurrency limit.
This makes it almost impossible to run other task instances in other DAGs unless the
large DAG is stopped.

In this PR, I added a column on the task instance table to know when the scheduler
examined the task instance. Once a dag has reached any of the concurrency limits, the value
of this column is updated and when any task instance has finished running,
the column is nullified so that the scheduler will examine the task instances again.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by ephraimbuddy** [comment]
Will add more tests later

File: airflow/logs/scheduler/latest [hunk 1]
```diff
@@ -0,0 +1 @@
+./logs/scheduler/2021-12-17
```
Comment (jedcunningham) [inline airflow/logs/scheduler/latest]: I assume this was accidental?

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -581,6 +593,13 @@ def _process_executor_events(self, session: Session = None) -> int:
                 self.log.info("Setting external_id for %s to %s", ti, info)
                 continue
 
+            # update the last_scheduling_decision for all task in this dag
+            session.query(TI).filter(TI.dag_id == ti.dag_id, TI.state == State.SCHEDULED,).update(
+                {
+                    TI.last_scheduling_decision: timezone.utcnow(),
```
Comment (jedcunningham) [inline airflow/jobs/scheduler_job.py]:
```suggestion
            # reset the last_scheduling_decision for all task in this dag
            session.query(TI).filter(TI.dag_id == ti.dag_id, TI.state == State.SCHEDULED).update(
                {
                    TI.last_scheduling_decision: None,
```

Don't we want to set them to null here? How does this interplay with the mini scheduler loop?

File: airflow/logs/scheduler/latest (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/logs/scheduler/latest]: Yeah!

**Comment by ephraimbuddy** [comment]
> Out of curiosity, is this a fix for newly introduced regression by one of recent commits or its been present in 2.x tree of Airflow since day one?

It has been present in 2.x. Using pools properly will help solve it to reasonable extent

File: tests/jobs/test_scheduler_job.py [hunk 3]
```diff
@@ -645,6 +645,89 @@ def test_find_executable_task_instances_in_default_pool(self, dag_maker):
...
+        self.scheduler_job = SchedulerJob(subdir=os.devnull)
+        res = self.scheduler_job._executable_task_instances_to_queued(max_tis=32, session=session)
+        assert 1 == len(res)
+        assert session.query(TaskInstance).filter(TaskInstance.state == TaskInstanceState.QUEUED).count() == 1
+        self.scheduler_job._executable_task_instances_to_queued(max_tis=32, session=session)
+        session.flush()
+        # assert last_scheduling_decision is updated
+        assert (
+            session.query(TaskInstance)
+            .filter(TaskInstance.state == State.SCHEDULED, TaskInstance.last_scheduling_decision.__ne__(None))
```
Comment (ephraimbuddy) [inline tests/jobs/test_scheduler_job.py]: Had to use `__ne__` because `is_not` was not working(I think cos of our sqlalchemy version)

Comment (kaxil) [reply tests/jobs/test_scheduler_job.py]: https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.ColumnOperators.__ne__

>In a column context, produces the clause a != b. If the target is None, produces a IS NOT NULL.

I am surprised it didn't work

Comment (ephraimbuddy) [reply tests/jobs/test_scheduler_job.py]: Maybe it's not available in SQLAlchemy==1.3.24 which we use

File: airflow/models/dagrun.py [hunk 4]
```diff
@@ -96,14 +94,6 @@ class DagRun(Base, LoggingMixin):
...
     last_scheduling_decision = Column(UtcDateTime)
     dag_hash = Column(String(32))
-    # Foreign key to LogFilename. DagRun rows created prior to this column's
-    # existence have this set to NULL. Later rows automatically populate this on
-    # insert to point to the latest LogFilename entry.
-    log_filename_id = Column(
-        Integer,
-        ForeignKey("log_filename.id", name="task_instance_log_filename_id_fkey", ondelete="NO ACTION"),
-        default=select([func.max(LogFilename.__table__.c.id)]),
-    )
```
Comment (ephraimbuddy) [inline airflow/models/dagrun.py:106]: surprised! not sure how I removed all these

File: tests/jobs/test_scheduler_job.py (hunk 3 above)
Comment (uranusjr) [reply tests/jobs/test_scheduler_job.py]: I think `isnot()` works in all 1.x versions.

Edit: I quickly searched and yes we do use `isnot()` in a couple of places. But not `is_not()`, that name is probably introduced only in 1.4.

File: airflow/jobs/scheduler_job.py [hunk 5]
```diff
@@ -395,6 +401,15 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
                         dag_id,
                         max_active_tasks_per_dag_limit,
                     )
+                    # reset the last_scheduling_decision for all task in this dag
```
Comment (ephraimbuddy) [inline airflow/jobs/scheduler_job.py]:
```suggestion
                    # update the last_scheduling_decision for all task instances in this dag
```

**Comment by uranusjr** [comment]
Are there scenarios where tasks in a DAG would have different `last_scheduling_decision`? I kind of feel it makes more sense to put this on DagRun instead.

File: airflow/jobs/scheduler_job.py [hunk 6]
```diff
@@ -289,7 +290,12 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
             .filter(not_(DM.is_paused))
             .filter(TI.state == TaskInstanceState.SCHEDULED)
             .options(selectinload('dag_model'))
-            .order_by(-TI.priority_weight, DR.execution_date)
+            .order_by(
+                -TI.priority_weight,
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job.py]:
```suggestion
                TI.priority_weight.desc(),
```

For consistency? Doesn’t really matter.

**Comment by ephraimbuddy** [comment]
> Are there scenarios where tasks in a DAG would have different `last_scheduling_decision`? I kind of feel it makes more sense to put this on DagRun instead.

There's `last_scheduling_decision` on DagRun. It helps on solving this kind of case for dagrun creation.

In this particular case, task instances of a DAG can have different values for last_scheduling_decision and some may not have at all. It's only task instances in SCHEDULE state that has their last_scheduling_decision updated when the DAG reaches the concurrency limit.

I have a simple DAGs to reproduce this behaviour and here is it:

1st DAG:
```python

from airflow import DAG
from datetime import datetime

dag = DAG(
    "airflow_bug",
    schedule_interval="0 1 * * *",
    start_date=datetime(2021, 1, 1),
    max_active_runs=1,
    concurrency=1,
)

for i in range(100):
    @dag.task(task_id=f'mytasrk{i}')
    def sleeping():
        import time
        time.sleep(60)
    sleeping()
```
2nd DAG:

```python
from airflow import DAG
from datetime import datetime

dag = DAG(
    "airflow_bug2",
    schedule_interval="0 1 * * *",
    start_date=datetime(2021, 1, 1),
    max_active_runs=1,
    concurrency=2,
)

for i in range(100):
    @dag.task(task_id=f'mytasrk{i}')
    def sleeping():
        import time
        time.sleep(60)
    sleeping()
```

When any one of these DAG reaches the concurrency limit, the other one will not run. 
To test this solution, you can duplicate the DAG further with different names and ensure the concurrency in one is not limiting the others. I got into a problem where some DAGs with older dagruns stops being picked. I have resolved it but will appreciate another eyes on that.

**Review by uranusjr — APPROVED** [review]

File: airflow/jobs/scheduler_job.py [hunk 7]
```diff
@@ -289,7 +290,12 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
             .filter(not_(DM.is_paused))
             .filter(TI.state == TaskInstanceState.SCHEDULED)
             .options(selectinload('dag_model'))
-            .order_by(-TI.priority_weight, DR.execution_date)
+            .order_by(
+                TI.priority_weight.desc(),
+                desc(case([(TI.last_scheduling_decision.is_(None), 1)], else_=0)),
+                TI.last_scheduling_decision.desc(),
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:304]: Please use `airflow.utils.sqlalchemy.nulls_first` here if we can.

It might not work out of the box for `DESC` sorted column.

File: airflow/jobs/scheduler_job.py [hunk 8]
```diff
@@ -395,6 +401,15 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
...
                         max_active_tasks_per_dag_limit,
                     )
+                    # update the last_scheduling_decision for all task instances in this dag
+                    session.query(TI).filter(
+                        TI.dag_id == task_instance.dag_id, TI.state == State.SCHEDULED
+                    ).update(
+                        {
+                            TI.last_scheduling_decision: timezone.utcnow(),
+                        },
+                        synchronize_session=False,
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:420]: This update is overly broad -- it is going to update _every single task for this DAG_, even completed ones.

**Review by ashb — DISMISSED** [review]
Your update query is going to (potentially) update hundreds of thousands of TaskInstance rows if run in a DB with a large amount of history.

File: airflow/migrations/versions/3457dcfe0528_add_last_scheduling_decision_column_to_.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+down_revision = '587bdf053233'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    """Apply add last_scheduling_decision column to taskinstance"""
+    with op.batch_alter_table('task_instance', schema=None) as batch_op:
+        batch_op.add_column(sa.Column('last_scheduling_decision', TIMESTAMP, nullable=True))
+        batch_op.create_index('idx_ti_last_scheduling_decision', ['last_scheduling_decision'], unique=False)
```
Comment (ashb) [inline airflow/migrations/versions/3457dcfe0528_add_last_scheduling_decision_column_to_.py]: This index should probably include at least the dag_id as well.

File: airflow/models/baseoperator.py [hunk 10]
```diff
@@ -666,7 +666,7 @@ def __init__(
             )
         self.weight_rule = weight_rule
         self.resources: Optional[Resources] = Resources(**resources) if resources else None
-        if task_concurrency and not max_active_tis_per_dag:
```
Comment (ashb) [inline airflow/models/baseoperator.py]: This change seems unrelated?

File: airflow/jobs/scheduler_job.py (hunk 7 above)
Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:304]: The nulls_first didn't work for other DBs except postgres. 

I'm preferring this PR https://github.com/apache/airflow/pull/19747 over this. What do you think?

Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:304]: Used nullsfirst for postgresql. I used it on the column directly instead of using `airflow.utils.sqlalchemy.nulls_first` since we are sure it's postgres

File: airflow/jobs/scheduler_job.py (hunk 8 above)
Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:420]: My comment here was wrong, it actually updates only those in scheduled state

File: airflow/jobs/scheduler_job.py [hunk 11]
```diff
@@ -581,6 +604,13 @@ def _process_executor_events(self, session: Session = None) -> int:
                 self.log.info("Setting external_id for %s to %s", ti, info)
                 continue
 
+            # reset the last_scheduling_decision for all task in this dag
```
Comment (ephraimbuddy) [inline airflow/jobs/scheduler_job.py]:
```suggestion
            # reset the last_scheduling_decision for all scheduled task in this dag
```

File: airflow/models/baseoperator.py (hunk 10 above)
Comment (ephraimbuddy) [reply airflow/models/baseoperator.py]: Not related but it seems like what we should do and might be ok in this PR

File: airflow/jobs/scheduler_job.py [hunk 12]
```diff
@@ -289,8 +290,21 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
             .filter(not_(DM.is_paused))
             .filter(TI.state == TaskInstanceState.SCHEDULED)
             .options(selectinload('dag_model'))
-            .order_by(-TI.priority_weight, DR.execution_date)
         )
+        if session.bind.dialect.name == "postgresql":
+            query = query.order_by(
+                TI.priority_weight.desc(),
+                TI.last_scheduling_decision.desc().nullsfirst(),
```
Comment (jedcunningham) [inline airflow/jobs/scheduler_job.py:297]: Don't we want to look in asc order with nullsfirst?

**Review by jedcunningham — COMMENTED** [review]
I'm kinda torn on the name `last_scheduling_decision` here. By the name, one could assume it'd be populated any time the scheduler considers the TI, not just when `max_active_tasks` is hit. It makes me wonder if we should populate that in more places?

Related, nulling it out after another task finishes also feels odd given the name of the column.

File: airflow/jobs/scheduler_job.py (hunk 12 above)
Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:297]: Using desc helps us go from the oldest date to recent. Using asc may make it difficult to reach a dag that first hit the max_active_runs

**Comment by ephraimbuddy** [comment]
> I'm kinda torn on the name `last_scheduling_decision` here. By the name, one could assume it'd be populated any time the scheduler considers the TI, not just when `max_active_tasks` is hit. It makes me wonder if we should populate that in more places?
> 
> Related, nulling it out after another task finishes also feels odd given the name of the column.

Makes sense, do you have suggestions on the name? Thinking about it too

**Comment by ashb** [comment]
I think we need to benchmark this approach and #19747 and see which one performs better under load

File: airflow/jobs/scheduler_job.py [hunk 13]
```diff
@@ -593,6 +616,13 @@ def _process_executor_events(self, session: Session = None) -> int:
                 self.log.info("Setting external_id for %s to %s", ti, info)
                 continue
 
+            # reset the last_scheduling_decision for all task instances in this dag
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:619]: Please change these comments to say why we need to reset it.

File: airflow/jobs/scheduler_job.py (hunk 12 above)
Comment (jedcunningham) [reply airflow/jobs/scheduler_job.py:297]: Yeah, I figured we wanted oldest first, which would be asc, no?

**Comment by jedcunningham** [comment]
Okay, I benchmarked 2.2.3, #19747 on latest main, and #20391 on latest main with:

- 100 DAGs, 10 runs each, 10 task per DAG for 10k total TIs
- 2 workers (1 CPU core each)
- 1 scheduler (1 CPU core)

It took each this long to get through them all:

* 2.2.3: 51 minutes
* 19747: 38 minutes
* 20391: 38 minutes

#19747 however does spew this into the scheduler logs ~16x per second (quite chatty!):

```
[2022-02-02 23:21:32,947] {scheduler_job.py:522} INFO - Setting the following tasks to queued state:
```

**Comment by ephraimbuddy** [comment]
Closing this in preference of https://github.com/apache/airflow/pull/19747
